std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,289 @@ unittest
                 2, "two",
                 ));
 }
+
+alias AllocateGC = Flag!"AllocateGC";
+
+private bool checkLength(Range1, Range2)(Range1 r1, Range2 r2)
+    if (isInputRange!Range1 &&
+        isInputRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
+    // If the two ranges aren't the same length, then they cannot
+    // possibly be permutations
+    static if (hasLength!(Range1) && hasLength!(Range2))
+    {
+        if (r1.length != r2.length)
+        {
+            return false;
+        }
+    }
+    else
+    {
+        // Custom walk length check to reduce the iterations as much as possible
+        while (!r1.empty)
+        {
+           if (r2.empty)
+           {
+              return false;
+           }
+
+           r1.popFront;
+           r2.popFront;
+        }
+
+        if (!r2.empty)
+        {
+            return false;
+        }
+    }
```

Comment:
If we're going through the trouble of making a special function to do this (especially if we want to make it public, and I think that's a good idea), we should make sure we don't walk the length of ranges with length. So if r1 has length, but r2 doesn't, or vice versa, only walk the length of one of them.